inductiva/_benchmarks/benchmark.py

```diff
@@ -0,0 +1,226 @@
+                for the simulations. These will update any existing parameters
+                with the same names.
+
+        Returns:
+            Self: The current instance for method chaining.
+        """
+        self.simulator = simulator or self.simulator
+        self.input_dir = input_dir or self.input_dir
+        self.on = on or self.on
+        self.kwargs = kwargs or self.kwargs
+        return self
+
+    def add_run(
+        self,
+        simulator: Optional[Simulator] = None,
+        input_dir: Optional[str] = None,
+        on: Optional[types.ComputationalResources] = None,
+        **kwargs,
+    ) -> Self:
+        """
+        Adds a simulation run to the benchmark.
+
+        Args:
+            simulator (Optional[Simulator]): The simulator to be used for this
+                run. If not provided, the previously set simulator will be used.
+            input_dir (Optional[str]): The directory containing input files for
+                the simulation. If not provided, the previously set input
+                directory will be used.
+            on (Optional[types.ComputationalResources]): The computational
+                resources to run the simulation on. If not provided, the
+                previously set resources will be used.
+            **kwargs: Additional keyword arguments for the simulator run. These
+                will overwrite any previously set parameters.
+
+        Returns:
+            Self: The current instance for method chaining.
+        """
+        self.runs.append((
+            simulator or self.simulator,
+            input_dir or self.input_dir,
+            on or self.on,
+            {
+                **self.kwargs,
+                **kwargs
+            },
+        ))
+        return self
+
+    def run(self, num_repeats: int = 2) -> Self:
+        """
+        Executes all added runs.
+
+        Each run is executed the specified number of times, and the collection
+        of runs is cleared afterwards.
+
+        Args:
+            num_repeats (int): The number of times to repeat each simulation
+            run (default is 2).
+
+        Returns:
+            Self: The current instance for method chaining.
+        """
+        with self:
+            for simulator, input_dir, machine_group, kwargs in self.runs:
+                machine_group.start(wait_on_pending_quota=True)
+                for _ in range(num_repeats):
+                    simulator.run(input_dir=input_dir,
+                                  on=machine_group,
+                                  **kwargs)
+            self.runs.clear()
+        return self
+
+    def wait(self) -> Self:
+        """
+        Waits for all running tasks to complete.
+
+        Returns:
+            Self: The current instance for method chaining.
+        """
+        tasks = self.get_tasks()
+        for task in tasks:
+            task.wait(download_std_on_completion=False)
+        return self
```

Comment:
I'll just reinforce my suggestion that all machine groups associated with the finished tasks should be terminated in this method.

Reply:
Yes, if a machine is terminated twice the API returns an error. But could that happen in this scenario?

Reply:
Why don't we just start all machine groups with a max idle time of 1 minute?

Reply:
Because that's user's decision -- the argument ```max_idle_time``` is passed in the ctor of ```MachineGroup```

Reply:
> Yes, if a machine is terminated twice the API returns an error. But could that happen in this scenario?

I'm thinking that users might be surprised by an unexpected exception raised when calling ```wait``` due to its side effect of invoking ```terminate```. Because, in contrast, calling ```run``` on the same machine only warns the user that the machine has already started. 

Maybe, it might be more appropriate to also pass a boolean parameter named ```terminate``` to the ```wait``` method (as well as to the```run``` method passing a ```start``` parameter) to provide better control over the method's behavior.

Reply:
> > Yes, if a machine is terminated twice the API returns an error. But could that happen in this scenario?
> 
> I'm thinking that users might be surprised by an unexpected exception raised when calling `wait` due to its side effect of invoking `terminate`. Because, in contrast, calling `run` on the same machine only warns the user that the machine has already started.
> 
> Maybe, it might be more appropriate to also pass a boolean parameter named `terminate` to the `wait` method (as well as to the`run` method passing a `start` parameter) to provide better control over the method's behavior.

I don't agree. The exception shouldn't happen unless the user tries to terminate the machine groups manually while the benchmark is running. The goal of introducing the termination is to avoid wasting resources. Most users won't be aware the machine groups will still be running after terminating the benchmark.

Reply:
Given the current interface, it would be weird to overwrite attributes of the machine group. But I understand the point. Alternatively we could provide the `terminate` method and print a message after starting the machine groups to warn the users that the machines will continue to run after the tasks are completed.

Reply:
> > > Yes, if a machine is terminated twice the API returns an error. But could that happen in this scenario?
> > 
> > 
> > I'm thinking that users might be surprised by an unexpected exception raised when calling `wait` due to its side effect of invoking `terminate`. Because, in contrast, calling `run` on the same machine only warns the user that the machine has already started.
> > Maybe, it might be more appropriate to also pass a boolean parameter named `terminate` to the `wait` method (as well as to the`run` method passing a `start` parameter) to provide better control over the method's behavior.
> 
> I don't agree. The exception shouldn't happen unless the user tries to terminate the machine groups manually while the benchmark is running. The goal of introducing the automatic termination is to avoid wasting resources. Most users won't be aware that the machine groups will still be running after terminating the benchmark.

I think the exception also happens when the user attempts to call the ```wait``` method twice, like this: ```benchmark.wait().wait()``` -- this is not a problem when calling ```task.wait().wait()```.

But, I agree that most users may not be aware of this and the machine groups should be automatically shut down.